### PR TITLE
[CSL-1037] Rewrite FollowTheSatoshi with StateT

### PR DESCRIPTION
**Pull request by int-index**
This is a rewrite of the `followTheSatoshiM' procedure.

* Now it doesn't rely on mutual recursion, so it's easier to understand.
* It uses more appropriate types (`CoinIndex` instead of `Coin`, `LocalSlotId` instead of `Int`).
* Additionally I documented the algorithm, with an example of execution.

### CODE REVIEW

File: lrc/Pos/Lrc/Fts.hs [hunk 1]
```diff
@@ -1,52 +1,250 @@
...
+coinIndexOffset c = over _CoinIndex (+ fromIntegral (unsafeGetCoin c))
+
+-- | Assign a local slot index to each value in a list, starting with
+-- @LocalSlotIndex 0@.
+assignToSlots :: [a] -> [(LocalSlotIndex, a)]
+assignToSlots = zip [0..]
+
+-- | Sort values by their local slot indices, then strip the indices.
+arrangeBySlots :: [(LocalSlotIndex, a)] -> [a]
+arrangeBySlots = map snd . sortOn fst
```
Comment (ChShersh) [inline lrc/Pos/Lrc/Fts.hs]: `sortOn fst` has worse performance than `sortBy (comparing fst)`

File: lrc/Pos/Lrc/Fts.hs [hunk 2]
```diff
@@ -1,52 +1,250 @@
...
+    -> Coin
+    -> m SlotLeaders
 followTheSatoshiM _ totalCoins
     | totalCoins == mkCoin 0 = error "followTheSatoshiM: nobody has any stake"
 followTheSatoshiM (SharedSeed seed) totalCoins = do
-    res <- findLeaders (sortOn fst $ zip coinIndices [1..]) (mkCoin 0)
-    pure . fromList . map fst . sortOn snd $ res
+    ftsState <- ftsStateInit <$> nextStakeholder
+    res <-
+        flip evalStateT ftsState $
```
Comment (ChShersh) [inline lrc/Pos/Lrc/Fts.hs]: If you don't like `flip evalStateT` there's [`evaluatingStateT`](https://github.com/serokell/universum/blob/70a9416b88624a388557b29a827182347d8654da/src/Monad/Trans.hs#L64) (about which you probably don't know yet) in `universum`

File: lrc/Pos/Lrc/Fts.hs [hunk 3]
```diff
@@ -1,52 +1,250 @@
...
+        findLeaders (sortOn snd $ assignToSlots coinIndices)
```
Comment (ChShersh) [inline lrc/Pos/Lrc/Fts.hs]: Same for `sortOn snd` here

**Comment by ChShersh** [comment]
I like this refactoring and enhanced documentation. Regarding commit: commits should start with issue tag, like `[CSL-228] Refactor FTS`. If issue is not created you can create new or use `SRK-8` for something general and minor (like refactoring).

File: lrc/Pos/Lrc/Fts.hs [hunk 4]
```diff
@@ -1,52 +1,248 @@
...
+--
+-- In other words, 'Coin' is a cardinal, whereas 'CoinIndex' is an ordinal.
+newtype CoinIndex = CoinIndex Integer
+    deriving (Show, Eq, Ord, Num, Enum)
+
+makePrisms ''CoinIndex
+
+-- | Offset a coin index by a given amount of coins.
+coinIndexOffset :: Coin -> CoinIndex -> CoinIndex
+coinIndexOffset c = over _CoinIndex (+ fromIntegral (unsafeGetCoin c))
```
Comment (pva701) [inline lrc/Pos/Lrc/Fts.hs]: @int-index are there any reasons not to use `unsafeAddCoin`? if there aren't, use it plz

Comment (int-index) [inline lrc/Pos/Lrc/Fts.hs]: The reason is that we're not adding `Coin`s, we're adding `CoinIndex`.

**Comment by neongreen** [comment]
@pva701 

> Also, when I rewrote FTS on StateT before, @neongreen rejected this changes. @neongreen why?

When?

**Comment by rockbmb** [comment]
@int-index CI doesn't seem to fail because of your commits, so when it does, I'm not opposed to merging this (although @pva701's concerns about performance should probably addressed).

If this ends up taking a few days, don't forget to merge `master` into this branch regularly.

**Comment by int-index** [comment]
@pva701 I doubt that `StateT` will be detrimental to performance. At worst it'll introduce a small constant factor, but the biggerst constant factor already comes from working in an abstract monad `m`. GHC has great optimization capabilities for concrete transformer stacks, so to address performance concerns we might want

1. compare lazy and strict `StateT`.
2. add `SPECIALIZE` pragmas for all instantiations of `m`.

Anyway, without a benchmark it's a waste of time.

**Comment by rockbmb** [comment]
@int-index 
> Anyway, without a benchmark it's a waste of time.

Then you should consider writing some, although I don't think this would have high priority.

File: lrc/Pos/Lrc/Fts.hs (hunk 4 above)
Comment (pva701) [inline lrc/Pos/Lrc/Fts.hs]: ok, got it

**Comment by pva701** [comment]
@neongreen I can't find this discussion in the #vatnik

**Comment by neongreen** [comment]
The build is failing because of HLint. @int-index please run `util-scripts/lint.sh` before making a pull request.

```
lrc/Pos/Lrc/Fts.hs:52:28: Warning: Use sortOn

Found:

  sortBy (comparing fst)

Why not:

  sortOn fst
```

```
lrc/Pos/Lrc/Fts.hs:216:29: Warning: Use sortOn

Found:

  sortBy (comparing snd)

Why not:

  sortOn snd
```

File: lrc/Pos/Lrc/Fts.hs (hunk 4 above)
Comment (neongreen) [inline lrc/Pos/Lrc/Fts.hs]: `over _CoinIndex (+ coinToInteger c)`

**Comment by rockbmb** [comment]
@int-index you know what, forget what I said. Unless this obviously breaks things in some way (we've seen that it doesn't), performance won't be a concern. It's not an ideal workflow, to completely disregard benches to changes in core code, but it's not like we haven't done that before. Carry on.

**Comment by neongreen** [comment]
I think that at this stage it doesn't make much sense to bother about benchmarks. Like, the most obvious problem here isn't `State`, it's that we make a call to the database at _each_ step instead of extracting several stakeholders at once :) As long as `State` doesn't cause a memory leak due to excessive laziness, or something, this code is fine.

So, the only check I would suggest is running it with a big utxo and checking whether it leaks or not. If it doesn't, it's fine.

**Comment by ChShersh** [comment]
@neongreen @int-index 
```
Found:

  sortBy (comparing snd)

Why not:

  sortOn snd
```
This should be disable in HLint. At least for this module. `sortOn` sometimes can be sooo less efficient than `sortBy . comparing`

**Comment by neongreen** [comment]
@ChShersh in Universum, `sortOn` is defined as `sortBy . comparing`. But even taking the `Data.List` version – how can it be much less efficient than `sortBy . comparing`?

image

**Comment by neongreen** [comment]
Oh, okay, I googled and apparently the overhead from decorating-and-undecorating can indeed be pretty big (7× in one case): https://ro-che.info/articles/2016-04-02-descending-sort-haskell. Anyway, this doesn't apply to our `sortOn`.

**Comment by int-index** [comment]
@ChShersh @neongreen I believe I addressed all of your comments. Can you approve? I'll rebase onto `master` as soon as it's fixed.

**Comment by neongreen** [comment]
> I'll rebase onto master

We usually merge (even when a clean rebase could be done). Also, master is broken anyway :trollface:

**Comment by ChShersh** [comment]
> @ChShersh in Universum, sortOn is defined as sortBy . comparing, so the hint is fine.

@neongreen well, since 0.4 it's not: https://github.com/serokell/universum/blob/70a9416b88624a388557b29a827182347d8654da/src/List.hs#L22

I've decided that having `sortOn` implementation with caching is better because we can benefit from it at some places. So it's better have two semantically different functions.

Another question: do we want to have some function like `sortOnButLikeSortBy = sortBy . comparing` to not write `sortBy . comparing` each time.

**Comment by neongreen** [comment]
> So it's better have two semantically different functions.

Makes sense.

> do we want to have some function like sortOnButLikeSortBy = sortBy . comparing to not write sortBy . comparing each time.

I do. Let's reexport `sortWith` from GHC.Exts in `universum` and then I'll change HLint.

**Comment by neongreen** [comment]
(no need to switch `sortOn` to `sortWith` in this PR, I'll do it across the codebase afterwards)

File: lrc/Pos/Lrc/Fts.hs [hunk 5]
```diff
@@ -1,52 +1,247 @@
...
+-- to particular stakeholders:
+--
+-- StakeholderId | Coin (amount) | Coin indices (inclusive range)
+-- --------------+---------------+-----------------------------
+--           xxx | 200           |   [0..199]
+--           yyy | 50            | [200..249]
+--           zzz | 80            | [250..329]
+--
+-- In other words, 'Coin' is a cardinal, whereas 'CoinIndex' is an ordinal.
+newtype CoinIndex = CoinIndex Integer
```
Comment (gromakovsky) [inline lrc/Pos/Lrc/Fts.hs]: Why is it `Integer` if we know that `maxBound :: Coin < maxBound :: Word64`?

Comment (int-index) [inline lrc/Pos/Lrc/Fts.hs]: Old code was using 1-based indexing, so the index of the last coin could not be represented. I changed it to 0-based, so now I can safely switch to `Word64` - good catch.

Comment (gromakovsky) [inline lrc/Pos/Lrc/Fts.hs]: Notice `<` (not `≤`). Max value of `Coin` is `45000000000000000`.

**Comment by gromakovsky** [comment]
> it's that we make a call to the database at each step instead of extracting several stakeholders at once :)

Are you sure that iterator doesn't load data in chunks?

**Comment by neongreen** [comment]
> Are you sure that iterator doesn't load data in chunks?

No. Does it load data in chunks?

**Comment by gromakovsky** [comment]
I don't know and don't want to investigate now, but I think chances that it does are high, because we are dealing with the database which aims at good performance :)

**Comment by gromakovsky** [comment]
Actually it should load all data in blocks and blocks should contain multiple entries.

**Comment by gromakovsky** [comment]
Anyway, don't want to think about it now.

**Comment by int-index** [comment]
Thank you for a thorough review, everyone!

**Pull request merged**